api/util/collections.hpp

```diff
@@ -0,0 +1,644 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+#ifndef UTIL_COLLECTIONS_HPP
+#define UTIL_COLLECTIONS_HPP
+
+#include <algorithm>
+#include <deque>
+#include <iosfwd>
+#include <list>
+#include <map>
+#include <numeric>
+#include <sstream>
+#include <type_traits>
+#include <vector>
+
+namespace collections {
+
+///
+/// Internal namespace consisting of collection traits
+///
+inline namespace traits {
+
+  template<typename>
+  struct is_sequence
+  : public std::false_type {};
+
+  template<typename T>
+  struct is_sequence<std::deque<T>>
+  : public std::true_type {};
+
+  template<typename T>
+  struct is_sequence<std::list<T>>
+  : public std::true_type {};
+
+  template<typename T>
+  struct is_sequence<std::vector<T>>
+  : public std::true_type {};
+
+  template<typename>
+  struct has_random_iterator
+  : public std::false_type {};
+
+  template<typename T>
+  struct has_random_iterator<std::deque<T>>
+  : public std::true_type {};
+
+  template<typename T>
+  struct has_random_iterator<std::vector<T>>
+  : public std::true_type {};
+
+} //< namespace traits
+
+///
+/// Internal namespace consisting of operation tag dispatchers
+///
+inline namespace operation_tags {
+
+  constexpr struct is_heap_tag {
+    constexpr is_heap_tag() noexcept {};
+  } is_heap;
+
+  constexpr struct is_sorted_tag {
+    constexpr is_sorted_tag() noexcept {};
+  } is_sorted;
+
+  constexpr struct make_heap_tag {
+    constexpr make_heap_tag() noexcept {};
+  } make_heap;
+
+  constexpr struct max_element_tag {
+    constexpr max_element_tag() noexcept {};
+  } max_element;
+
+  constexpr struct min_element_tag {
+    constexpr min_element_tag() noexcept {};
+  } min_element;
+
+  constexpr struct next_permutation_tag {
+    constexpr next_permutation_tag() noexcept {};
+  } next_permutation;
+
+  constexpr struct prev_permutation_tag {
+    constexpr prev_permutation_tag() noexcept {};
+  } prev_permutation;
+
+  constexpr struct sort_tag {
+    constexpr sort_tag() noexcept {};
+  } sort;
+
+  constexpr struct stable_sort_tag {
+    constexpr stable_sort_tag() noexcept {};
+  } stable_sort;
+
+  constexpr struct sum_tag {
+    constexpr sum_tag() noexcept {};
+  } sum;
+
+  constexpr struct unique_tag {
+    constexpr unique_tag() noexcept {};
+  } unique;
+
+} //< namespace operation_tags
+
+///
+/// Operator to check if the specified sequence is heap
+///
+/// @param data_set
+///   The specified sequence to check if it's a heap
+///
+/// @return true if it's heap, false otherwise
+///
+template<typename T,
+         typename = std::enable_if_t<
+                    has_random_iterator<
+                    std::remove_cv_t<std::remove_reference_t<T>>>::value>>
```

Comment:
Have you considered to use `std::decay_t` instead of `std::remove_cv_t` and `std::remove_reference_t`?